src/pybind/mgr/cephadm/serve.py

```diff
@@ -993,13 +986,24 @@ def _run_cephadm(self,
                 try:
                     out, err, code = remoto.process.check(
                         conn,
-                        [python, '-u'],
-                        stdin=script.encode('utf-8'))
+                        [python, self.mgr.cephadm_binary_path] + final_args,
+                        stdin=stdin.encode('utf-8') if stdin is not None else None)
+                    if code == 2:
+                        out_ls, err_ls, code_ls = remoto.process.check(
+                            conn, ['ls', self.mgr.cephadm_binary_path])
+                        if code_ls == 2:
```

Comment:
```suggestion
                        if code_ls != 0:
```
(nit) similar, but maybe any non-zero errcode?

Might also be good to extract this into a dedicated `cephadm_binary_exists` function?